src/workerd/api/pyodide/pyodide.h

```diff
@@ -65,12 +66,71 @@ class PyodideMetadataReader : public jsg::Object {
   }
 };
 
+// A loaded bundle of artifacts for a particular script id. It can also contain V8 version and
+// CPU architecture-specific artifacts. The logic for loading these is in getArtifacts.
+class ArtifactBundler : public jsg::Object {
+public:
+  ArtifactBundler(kj::Maybe<kj::Array<kj::byte>> existingSnapshot,
+      kj::Function<kj::Promise<bool>(kj::Array<kj::byte> snapshot)> uploadMemorySnapshotCb)
+      : existingSnapshot(kj::mv(existingSnapshot)),
+        uploadMemorySnapshotCb(kj::mv(uploadMemorySnapshotCb)),
+        hasUploaded(false) {};
+
+  ArtifactBundler()
+      : existingSnapshot(kj::none),
+        uploadMemorySnapshotCb(kj::none),
+        hasUploaded(false) {};
+
+  jsg::Promise<bool> uploadMemorySnapshot(jsg::Lock& js, kj::Array<kj::byte> snapshot) {
+    // Prevent multiple uploads.
+    JSG_REQUIRE(!hasUploaded, TypeError,
+        "This ArtifactBundler has already uploaded a memory snapshot");
+
+    KJ_IF_SOME(cb, uploadMemorySnapshotCb) {
+      hasUploaded = true;
+      auto& context = IoContext::current();
+      return context.awaitIo(cb(kj::mv(snapshot)));
+    } else {
+      JSG_FAIL_REQUIRE(TypeError, "ArtifactBundler is disabled");
+    }
+  };
+
+  jsg::Optional<kj::ArrayPtr<kj::byte>> getMemorySnapshot(jsg::Lock& js) {
+    KJ_IF_SOME(val, existingSnapshot) {
+      return kj::ArrayPtr<kj::byte>(val);
+    }
+    return kj::none;
+  }
+
+  bool isDisabled() {
+    return uploadMemorySnapshotCb == kj::none;
+  }
+
+  static jsg::Ref<workerd::api::pyodide::ArtifactBundler> makeDisabledBundler() {
+    return jsg::alloc<workerd::api::pyodide::ArtifactBundler>();
+  }
+
+  JSG_RESOURCE_TYPE(ArtifactBundler) {
+    JSG_METHOD(uploadMemorySnapshot);
+    JSG_METHOD(getMemorySnapshot);
+    JSG_METHOD(isDisabled);
+  }
```

Comment:
Non-blocking: since this is holding onto memory (even if not v8 heap memory) ... it's a good idea to implement the new memory tracking mechanism.

```
void visitForMemoryInfo(jsg::MemoryTracker& tracker) const {
  KJ_IF_MAYBE(snapshot, existingSnapshot) {
    tracker.trackFieldWithSize("snapshot", snapshot.size());
  }
}
```